Python/chapter02/2.1 - Remove Dups/miguel_2.1_sol.py

```diff
@@ -0,0 +1,165 @@
+"""
+Python version 3.7.0
+2.1 - Remove Dups
+Write code to remove duplicates
+from an unsorted linked list.
+How would you solve this if a temporary
+buffer is not allowed?
+"""
+import unittest
+from typing import List
+
+
+class Node:
+    def __init__(self, d: int):
+        self.data = d
+        self.next = None
+
+
+class LinkedList:
+    def __init__(self):
+        self.head = None
+
+    def append_to_tail(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            return
+        end = Node(d)
+        n = self.head
+        while n.next is not None:
+            n = n.next
+        n.next = end
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other):
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def build_linked_list(numbers: List[int]) -> LinkedList:
+    ll = LinkedList()
+    for num in numbers:
+        ll.append_to_tail(num)
+    return ll
+
+
+def remove_dups(ll: LinkedList) -> LinkedList:
+    """
+    remove_dups will remove duplicates from the
+    input linked list ll. A temporary buffer
+    is used.
+    Runtime: O(N)
+    Space Complexity: O(N)
+    :param ll: a linked list
+    :return: a linked list without duplicates
+    """
+    n = ll.head
+    unique_vals = set()
+    while n.next is not None:
+        unique_vals.add(n.data)
+        n = n.next
+    unique_vals.add(n.data)
+    return build_linked_list(list(unique_vals))
```

Comment:
what is the complexity of this operation?

Reply:
runtime for list(unique_vals) is O(N)
then to build a linked list from that:  O(N)

Reply:
can you prove the latter?

Reply:
Whoops, I overlooked the append_to_tail runtime, which is O(N).  So, this means that to build a linked list from 'numbers' using append_to_tail will be O(N^2).

I'm thinking I can loop from the end of numbers and append to head, which will be O(1) insertion.

Or, I can maintain a 'tail' pointer in the linked list to ensure O(1) insertion.

Reply:
either work, although the latter might be a bit more pythonic and simpler to understand.